lib/gis/omp_threads.c

```diff
@@ -0,0 +1,44 @@
+#if defined(_OPENMP)
+#include <omp.h>
+#endif
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <grass/gis.h>
+#include <grass/glocale.h>
+
+/*! \brief A helper function to setup the number of threads for C modules
+   supported by OpenMP
+   \param nprocs the number of threads specified by the user
+   \return the number of threads set up for OpenMP parallel computing
+ */
+
+int G_setup_threads(char *nprocs)
+{
+    int threads = atoi(nprocs);
+#if defined(_OPENMP)
+    int max_cores = omp_get_num_procs();
+    if (threads > max_cores) {
+        G_warning(
+            _("The number of threads specified is greater than the "
+              "number of processors available (%d). The number of threads "
+              "will be set to %d."),
+            max_cores, max_cores);
+        threads = max_cores;
```

Comment:
Is oversubscription so bad that we have to prevent it? The program will not fail.

Reply:
Is your idea to pass something like 1000 to get the maximum number of threads available? With a nice function like this, we could have `max` or `auto` as a value, but `0` does that already with the current implementation, no? (Unlike `0`, `auto`/`max`/... would require loosening the parser rules from int to string, so that makes `0` appealing.)

Reply:
> Is oversubscription so bad that we have to prevent it? The program will not fail.

Using threads > cores could benefit an I/O-bounded program because it has to wait for the OS or the internet to respond. Pipelining will handle threads' waiting time.

For compute-bounded grogram, threads > cores cannot get benefits but cause some overheads, such as context switching. More memory is allocated, and it is wasted. If a thread finishes slower than others and synchronization must be done in a loop, others must wait for it. Though the program may not fail, it degrades the performance.

This implementation is more conservative. We can let a user use more threads. It depends on whether our audiences are familiar with multithreading. Users may set 100 threads for every tool because they believe more is better. They always get overhead, but they don't know.

It also depends on whether GRASS GIS wants to let users manage their programs (flexible) or not (conservative). I believe you have a better idea about this, so let me know what you think.

Reply:
@HuidaeCho Do you have any suggestions here?

Reply:
Threads are not cores. Performance cores support hyper-threading, but efficiency cores don't, so I would rename all *cores to *threads.

Reply:
As far as I know, `Processes == Cores != Threads`, so that message is misleading. The number of threads (in OpenMP terms) *can be* greater than the number of processes (in MPI terms).

Reply:
@wenzeslaus I think we could remove the thread limitation here but keep the warning, and I will modify the warning message based on @HuidaeCho's comments.

Reply:
> > Is oversubscription so bad that we have to prevent it? The program will not fail.
> 
> Using threads > cores could benefit an I/O-bounded program because it has to wait for the OS or the internet to respond. Pipelining will handle threads' waiting time.
> 
> For compute-bounded grogram, threads > cores cannot get benefits but cause some overheads, such as context switching. More memory is allocated, and it is wasted. If a thread finishes slower than others and synchronization must be done in a loop, others must wait for it. Though the program may not fail, it degrades the performance.
> 
> This implementation is more conservative. We can let a user use more threads. It depends on whether our audiences are familiar with multithreading. Users may set 100 threads for every tool because they believe more is better. They always get overhead, but they don't know.
> 
> It also depends on whether GRASS GIS wants to let users manage their programs (flexible) or not (conservative). I believe you have a better idea about this, so let me know what you think.

It would be great to add a reference for this argument in the code and message. Also, explain it if it's greater than the max in the message (or better if we already know I/O-bound vs. compute-bound per module, give a better advice), but still not limiting it (if still useful)?